### PR TITLE
[Chores] Automated API doc generation and added packages/samples filter

### PR DESCRIPTION
**Pull request by illiakovalenko**
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated

## Description / Motivation
* Added "generate API docs" pipeline step:
  * Generates API docs only for packages that were affected by the changes in PR
  * Commits and pushes them using '[skip ci]' identifier
  * Will be triggered ONLY for _dev_|_release_ branches, since doc generation provides many changes by updating hash numbers for all of the files in a specific package, in this case, each PR will include many changed files that will complicate PR for review
* Introduced **lint-samples** script that triggers lint step ONLY for the samples that were directly affected by the changes in PR
* Added lerna **--since** flag for _generate-docs_ and _lint-packages_ to include ONLY packages that were directly affected by the changes in PR
* Our internal release-prep script needs to be updated to replace _"--since=origin/dev" by "--since=origin/release/x"_

## Testing Details

- [ ] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -87,21 +87,37 @@ steps:
...
+  # LERNA VERSION
+  # * Increment pre-release version
+  #     * Force version update even if lerna doesn't detect changes
+  #     * Include [skip ci] in commit message to stop pipeline from triggering again when lerna pushes
+  # * Only run for 'dev' and 'release' branches
+  #
+- script: |
     $(npm bin)/lerna version prerelease --preid canary --force-publish --message "version %s [skip ci]" --yes
   displayName: 'pre-release version update'
   condition: and(succeeded(), eq(variables.shouldPublish, true))
```
Comment (ambrauer) [inline azure-pipelines.yml:123]: Can get rid of the GITHUB_PAT_ENV from this step's envs (below)

File: scripts/lint-samples.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+const { execSync } = require('child_process');
+const samples = require('./samples.json');
+
+/**
+ * Start linting process only for the samples that were affected by the new changes in:
+ * - create-sitecore-jss/src/templates/**
+ * - create-sitecore-jss/src/initializers/**
+ */
+
+const affectedTemplates = execSync('git diff --name-only packages/create-sitecore-jss', {
```
Comment (ambrauer) [inline scripts/lint-samples.js:10]: Do we need to explicitly specify the target branch here?

File: scripts/lint-samples.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  encoding: 'utf-8',
+})
+  .split('\n')
+  .map((filepath) => {
+    // Extracting template names from the filepath
+    const template = filepath.match(/(templates|initializers)\/([^\/]*)/);
```
Comment (ambrauer) [inline scripts/lint-samples.js:16]: Looks like this approach would miss changes made in src/common or src/<root> files. Though, I don't know if that's a huge concern? Those changes would be caught in our separate automated tests suite.

File: scripts/lint-samples.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  console.log('No modified templates to lint');
+
+  return;
+}
+
+const affectedSamples = samples
+  .filter((sample) => sample.initializers.some((template) => affectedTemplates.includes(template)))
+  .map(
+    (sample) =>
+      `sample-${sample.args.appName}-${sample.args.fetchWith || ''}-${sample.args.prerender || ''}`
```
Comment (ambrauer) [inline scripts/lint-samples.js:33]: This was done to match scripts/scaffold-samples.js I see. Can we create a helper function and re-use this logic in both places?

**Review by ambrauer — CHANGES_REQUESTED** [review]
Looks great, love the optimization to samples linting. See some comments.

File: scripts/lint-samples.js (hunk 3 above)
Comment (illiakovalenko) [reply scripts/lint-samples.js:16]: You are right, I think that we need to consider more specific changes, that are directly related to the template
rest of them will be caught

File: azure-pipelines.yml (hunk 1 above)
Comment (illiakovalenko) [reply azure-pipelines.yml:123]: Yeah, it's not needed here

File: scripts/lint-samples.js (hunk 4 above)
Comment (illiakovalenko) [reply scripts/lint-samples.js]: Good call, added _utils.js_ file

File: scripts/lint-samples.js (hunk 2 above)
Comment (illiakovalenko) [reply scripts/lint-samples.js]: Valid point, I also updated our internal script to replace _origin/dev_ during the release prep

**Pull request merged**